doc/developer-guides/hld/sriov-hld.rst

```diff
@@ -0,0 +1,303 @@
+device *sysfs* node. Hypervisor intercepts all SR-IOV capability
+accesses and check *VF\_ENABLE* state, if *VF\_ENABLE* is clear,
+hypervisor makes VF virtual devices invisible from SOS that all accesses
+for VF will return 0xFFFFFFFF as an error, VF physical devices are
+removed within 1s of *VF\_ENABLE* is clear.
+
+SR-IOV VF Assignment Policy
+---------------------------
+
+.. figure:: images/sriov-image8.png
+   :align: center
+   :width: 900px
+   :name: SR-IOV-vf-assignment
+
+   SR-IOV VF Assignment
+
+1. All SR-IOV PF devices are managed by SOS.
+
+2. Currently, SR-IOV PF can’t passthrough to UOS.
+
+3. All VFs can be passthrough to UOS, but we do not recommend
+   passthrough to high privilege VMs, because PF device may impact
+   assigned VFs including functionality and stability.
+
+SR-IOV Usage Guide In ACRN
+--------------------------
+
+We use Intel 82576 NIC as an example in here and only support LaaG
+(Linux as a Guest).
+
+1. Please ensure the 82576 PF and VF drivers are compiled into SOS
+   Kernel and VF driver is compiled into UOS Kernel (Set *CONFIG\_IGB=y*
+   and *CONFIG\_IGBVF=y* in Kernel Config)
+
+2. When the SOS boot up, “\ *lspci -v*\ ” command can show Intel 82576
+   NIC devices have SR-IOV capability and their PF drivers are igb.
+
+.. figure:: images/sriov-image9.png
+   :align: center
+   :width: 900px
+   :name: 82576-pf
+
+   82576 SR-IOV PF devices
+
+1. Input “\ *echo n > /sys/class/net/enp109s0f0/device/sriov\_numvfs*\ ”
+   command in SOS to enable n VF devices for the first PF
+   device(\ *enp109s0f0)*. The number *n* can’t more than *TotalVFs*
+   that comes from the return value of command
+
+   *“cat /sys/class/net/enp109s0f0/device/sriov\_totalvfs”*. Here we use
+   *n = 2* for the example.
+
+.. figure:: images/sriov-image10.png
+   :align: center
+   :width: 900px
+   :name: 82576-vf
+
+   82576 SR-IOV VF devices
+
+.. figure:: images/sriov-image11.png
+   :align: center
+   :width: 900px
+   :name: 82576-vf-nic
+
+   82576 SR-IOV VF NIC
+
+1. Passthrough a SR-IOV VF device to guest
+
+   a. Unbind igbvf driver in SOS
+
+      i.   *modprobe pci\_stub*
+
+      ii.  *echo "8086 10ca" > /sys/bus/pci/drivers/pci-stub/new\_id*
+
+      iii. *echo "0000:6d:10.0" >
+           /sys/bus/pci/devices/0000:6d:10.0/driver/unbind*
+
+      iv.  *echo "0000:6d:10.0" >
+           /sys/bus/pci/drivers/pci-stub/bind*
+
+   b. Add SR-IOV VF device parameter (“*-s X, passthru,6d/10/0*\ ”) in
+      launch UOS script
+
+.. figure:: images/sriov-image12.png
+   :align: center
+   :width: 900px
+   :name: 82576-nic-passthru
+
+   Configure 82576 NIC as Passthrough Device
+
+a. Boot UOS
+
+SR-IOV Limitations In ACRN
+--------------------------
+
+1. The SR-IOV migration feature is not supported.
+
+2. If the number of a PF device and all its VF devices exceeds the
+   maximum number of devices supported by the platform, the PF device
+   will not be dropped.
```

Comment:
I don't quite understand this statement (and it's implications). Can you expand?

Reply:
Update it as below.

   If one SR-IOV PF device is detected during enumeration phase, but there is no
   enough room for its total VF devices, the PF device will be dropped. The platform
   uses *MAX_PCI_DEV_NUM* ACRN configuration to support the maximum number of PCI
   devices, please make sure *MAX_PCI_DEV_NUM* is more than the number of all PCI
   devices including total SR-IOV VF devices.